### PR TITLE
feat: Feature Flag Component + Simplify the traversal for composability

### LINKED ISSUE #5480: core: add a FeatureFlagged component
Giving this a new home since https://github.com/backstage/backstage/issues/5394 is closed :grin:

## Feature Suggestion

Looking for feedback on the idea to add a single `FeatureFlagged` component for use in the app, that'll both take care of acting as a switch but also register the feature flag. It would be implemented to work with the composability API, and simply act as a switch for whether its children should be renderer. Basically this:

```tsx
const routes = (
  <FlatRoutes>
...
    <FeatureFlagged flag="show-welcome-page">
      <Route path="/welcome" element={<WelcomePage />} />
    <FeatureFlagged>
...
  <FlatRoutes>
)
```

Where the `Route` works as expected, is discovered in the app etc, and the `"show-welcome-page"` flag is both automatically registered as well as used to toggle whether the welcome page should be renderer. I'm not sure if that exact implementation is possible, but something of that sort x)

This'd lift the feature flag registration from plugins to the app, which we can ofc debate a bit ;)

## Possible Implementation

TBD, looking at the idea first x)

## Context

https://github.com/backstage/backstage/issues/5394

### PR DESCRIPTION
**Pull request by benjdlambert**
## Hey, I just made a Pull Request!

Fixes #5480 

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by Fox32** [comment]
When adding feature flags, I sometimes replace an existing component with a new one. Therefore it would be neat if the `<FeatureFlagged>` component has a way to negate the condition or an else branch ("run this instead if the feature flag is disabled).

**Comment by Fox32** [comment]
For example, we have this code right now:

```ts
    {isFeatureFlagActive('search-2-0') ? (
      <Route path="/search" element={<SearchPage />}>
        {searchPage}
      </Route>
    ) : (
      <Route path="/search" element={<SearchPage />} />
    )}
```

Where `isFeatureFlagActive` is a helper we build ourselfs.

**Comment by Rugvip** [comment]
One though I had was to just add an `invert` prop that flips things around so that things are rendered when the flag is off

So this:

```tsx
<FeatureFlagged flag="search-2-0">
  <Route path="/search" element={<SearchPage />}>
    {searchPage}
  </Route>
</FeatureFlagged>
<FeatureFlagged flag="search-2-0" invert>
  <Route path="/search" element={<SearchPage />} />
</FeatureFlagged>
```

It's a bit more verbose but it feels like we'll also want to be able to invert it in general anyway. Think that would be good enough @Fox32?

**Comment by Fox32** [comment]
> Think that would be good enough @Fox32?

I think that is sufficient. 

Another option that may reads a bit more fluid could be an `unless` prop (copied from handlebars). But that would require a bit more complex typing on the props to make sure that you either set `if` or `unless`, but not both.

**Comment by benjdlambert** [comment]
@Fox32 - had a little free time today, so had a stab at trying to include this functionality here part of this PR. Open to suggestions for the naming, but thought that we could try and condense this into a single prop with union types rather than prop which is acts an an operator like the `inverse` prop. We also thought a little bit of something like `flag="!my-flag"` which we can typecheck, but it's a little unintuitive when it comes to typescript tbh. 

This is what it currently looks like:


```tsx
<FeatureFlagged with="my-flag-that-is-enabled">
  <marquee>DJ Caspa</marquee>
</FeatureFlagged>
```

And then to render when the flag is disabled

```tsx
<FeatureFlagged without="my-flag-that-is-disabled">
  <marquee>Slide to the left</marquee>
</FeatureFlagged>
```

I think that you're right that it would be nicer to condense, kind of stayed away from the `unless` conjunctive here as it's always something that makes me think super hard about what's happening, and I think `without={flagName}` is maybe a little clearer and easier to understand? 

Thoughts?

**Comment by Fox32** [comment]
> Thoughts?

Looks good 👍

**Comment by freben** [comment]
Yeah this reads better than `invert`.

Potential drawback that it is different from prior art (like handlebars and our own optional composability components). Bit no biggie

File: packages/app/src/components/catalog/EntityPage.tsx [hunk 1]
```diff
@@ -282,9 +283,11 @@ const serviceEntityPage = (
       {overviewContent}
     </EntityLayout.Route>
 
-    <EntityLayout.Route path="/ci-cd" title="CI/CD">
-      {cicdContent}
-    </EntityLayout.Route>
+    <FeatureFlagged with="show-graphiql-page">
```
Comment (freben) [inline packages/app/src/components/catalog/EntityPage.tsx:286]: remove!

File: packages/app/src/components/catalog/EntityPage.tsx [hunk 2]
```diff
@@ -303,9 +306,11 @@ const serviceEntityPage = (
 
     <EntityLayout.Route path="/dependencies" title="Dependencies">
       <Grid container spacing={3} alignItems="stretch">
-        <Grid item md={6}>
-          <EntityDependsOnComponentsCard variant="gridItem" />
-        </Grid>
+        <FeatureFlagged with="show-graphiql-page">
```
Comment (freben) [inline packages/app/src/components/catalog/EntityPage.tsx:309]: remove!

File: packages/core-app-api/api-report.md [hunk 3]
```diff
@@ -234,6 +234,18 @@ export type ErrorBoundaryFallbackProps = {
     resetError: () => void;
 };
 
+// @public (undocumented)
+export const FeatureFlagged: (props: FeatureFlaggedProps) => JSX.Element | null;
+
+// @public (undocumented)
+export type FeatureFlaggedProps = {
+    children: JSX.Element | null;
```
Comment (freben) [inline packages/core-app-api/api-report.md:242]: Should this be ReactNode instead?

**Review by Rugvip — COMMENTED** [review]
I added docs, but I also switched the component data check from `Boolean(data)` to `data !== undefined`. I can see that filtering out falsy values could be quite unexpected and cause trouble in the future.

**Comment by Rugvip** [comment]
Ended up extracting `ElementCollection` to an interface which is now exported, which I think is good form. It also makes it tracked in API reports.

**Review by freben — APPROVED** [review]

**Pull request merged**